Implemented addAttachment method to TransportBuilder

dudzio12 opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
Implemented `addAttachment` method to `TransportBuilder` class which allows developers to add attachment to email messages.

### Fixed Issues (if relevant)
1. Fixes magento/magento2#33261

### Manual testing scenarios (*)
1. Tested transportBuilder without `addAttachment` usage.
2. Tested transportBuilder with `addAttachment` usage.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

dudzio12 committed:
Added addAttachment method to transport builder

danielrussob commented:
@magento give me test instance

danielrussob commented on lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php:
```diff
@@ -442,4 +460,28 @@ private function addAddressByType(string $addressType, $email, ?string $name = n
+     *
+     * @return $this
+     */
+    public function addAttachment(?string $content, ?string $fileName, ?string $fileType)
```
Hello @dudzio12 ,
thank you for your contribution!

According to: https://github.com/laminas/laminas-mime/blob/2.9.x/src/Part.php @ public function setContent($content)

if (! is_string($content) && ! is_resource($content)) it will throw an Exception\InvalidArgumentException, you should avoid ?string $content in parameter and add Exception\InvalidArgumentException in docs

also

according to https://github.com/laminas/laminas-mime/blob/2.9.x/src/Part.php @ public function setType($type = Mime::TYPE_OCTETSTREAM) could be a good idea set $fileType as default Mime::TYPE_OCTETSTREAM to avoid breaking changes after an update of laminas

dudzio12 replied:
Thanks for diving into the topic, I willl work on that.

dudzio12 replied:
Added `@throws InvalidArgumentException` annotation and changed argument type from `?string` to `string` for first two arguments.  I have also set `Mime::TYPE_OCTETSTREAM` as default value `$fileType` argument.

danielrussob commented:
@magento run all tests

aleron75 commented:
Hello @dudzio12!

First of all, thank you for your contribution.

Due to [Magento Definition of Done](https://devdocs.magento.com/contributor-guide/contributing_dod.html#automated-tests), all code must be covered by tests.

For this specific case, do you think it would be feasible to extend `\Magento\Framework\Mail\TransportBuilderTest` to add a test case that covers this new feature?

My best,
Alessandro

aleron75 commented on lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php:
```diff
@@ -25,6 +25,8 @@
 use Magento\Framework\Phrase;
+use Zend\Mime\Mime;
+use Zend\Mime\PartFactory;
 
```
If  I am not wrong, on Magento v2.4-develop the reference to `Zend` namespace should be changed to `Laminas`.

dudzio12 replied:
Thanks fo pointing that, I will change that in the next commit and prepare tests.

dudzio12 replied:
Changed `Zend\Mime\Mime` to `Laminas\Mime\Mime`. Removed `Zend\Mime\PartFactory` from constructor, since generated Factory class is not valid, I've used standard object initialization instead - there is no real need for Factory here.

dudzio12 committed:
Covered method with unit test

dudzio12 commented:
@magento run all tests

danielrussob commented:
@magento run all tests

danielrussob commented:
@magento run Integration Tests

danielrussob commented:
@magento run Sample Data Tests CE

aleron75 commented:
The semantic version test fails because new property & method has been introduced and it's correct, the QA team will take it into consideration when packaging the release.

We re-run the other failing tests because they seem flaky.

dudzio12 committed:
Fixed issue with attachment Transport port not resetting when using inside loop

dudzio12 committed:
Changed integration test

dudzio12 commented:
I've fixed issue when using TransportBuilder inside loop. It didn't reset attachment part, so on every next mail there were duplicated attachments.

I also changed integration test to implement new method.

dudzio12 commented:
@magento run all tests

dudzio12 committed:
Fixed code formatting according to coding standards

dudzio12 commented:
@magento run Integration Tests

dudzio12 commented:
@magento run all tests

danielrussob commented:
@magento run all tests

hostep commented:
Hi folks, what would be the status here?

Would be nice to have an easy way to add attachments to emails in core Magento 🙂 

/cc @ishakhsuvarov, @sidolov

dudzio12 committed:
Removed empty line

dudzio12 commented:
Branch was updated

dudzio12 commented:
@magento run all tests

dudzio12 commented:
I'm not sure what we can do about this:

```
PHP Code Mess has found error(s): /var/www/html/lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php:38 TooManyFields The class TransportBuilder has 16 fields. Consider redesigning TransportBuilder to keep the number of fields under 15. Failed asserting that 2 matches expected 0.
```

dudzio12 committed:
Merge branch 'magento:2.4-develop' into email-transport-builder-attachments

dudzio12 committed:
Merge branch '2.4-develop' into email-transport-builder-attachments

engcom-Hotel committed:
Merge branch '2.4-develop' into email-transport-builder-attachments

engcom-Hotel commented:
@magento run all tests

engcom-Hotel commented on lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php:
```diff
@@ -103,6 +105,11 @@ class TransportBuilder
+    /**
+     * @var array
+     */
+    protected $attachments = [];
```
I think we can make this `private`.

engcom-Hotel commented on lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php:
```diff
@@ -342,6 +349,29 @@ public function getTransport()
+     * @return $this
+     * @throws InvalidArgumentException
+     */
+    public function addAttachment(string $content, string $fileName, ?string $fileType = Mime::TYPE_OCTETSTREAM)
```
Please specify the return type of this method

engcom-Hotel commented:
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Integration Tests, Static Tests, WebAPI Tests

engcom-Hotel reviewed (changes requested):
Hello @dudzio12,

Thanks for the contribution!

Please have a look into the below review points.

Thanks

engcom-Hotel commented:
Hello @dudzio12,

Have you got a chance to look into the above [review comments](https://github.com/magento/magento2/pull/33262#pullrequestreview-1913959662)?

Thanks

engcom-Hotel committed:
Merge branch '2.4-develop' into email-transport-builder-attachments

engcom-Hotel commented:
We are picking this PR for further development

glo24157 committed:
Added return type to addAttachment method and make variable private

engcom-Hotel commented:
@magento run all tests

hostep commented on lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php:
```diff
@@ -27,6 +27,7 @@
 use Magento\Framework\Mail\TransportInterfaceFactory;
 use Magento\Framework\ObjectManagerInterface;
 use Magento\Framework\Phrase;
+use tests\verification\Tests\ActionGroupWithReturnGenerationTest;
```
@glo24157: what's this line about? This feels like a mistake?

engcom-Hotel replied:
Hello @hostep I have resolved this, please review it again.

hostep replied:
Thanks!

glo24157 committed:
Removed unused included file

engcom-Hotel commented:
@magento run all tests

engcom-Hotel committed:
Merge branch '2.4-develop' into email-transport-builder-attachments

glo24157 committed:
Fix static test failures

engcom-Hotel commented:
@magento run all tests

engcom-Hotel commented:
@magento run Functional Tests B2B, Functional Tests EE, Functional Tests CE

engcom-Hotel committed:
Fix integration test failures

engcom-Hotel commented:
@magento run Integration Tests, Static Tests

engcom-Hotel committed:
Merge branch '2.4-develop' into email-transport-builder-attachments

engcom-Hotel committed:
Fix static test failures

engcom-Hotel commented:
@magento run Integration Tests, Static Tests

engcom-Hotel committed:
Rollback extension_attribute.pdf

engcom-Hotel commented:
@magento run Integration Tests

dudzio12 commented:
I'll dig in this topic later this week and run tests locally.

engcom-Hotel commented:
@magento run all tests

engcom-Hotel commented:
@magento run Functional Tests B2B, Functional Tests EE, Functional Tests CE, Integration Tests

engcom-Hotel commented:
@magento run Functional Tests B2B, Functional Tests EE, Functional Tests CE, Integration Tests

engcom-Hotel commented:
Failed tests seems flaky to me hence moving the PR for "Review"

engcom-Hotel commented:
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

engcom-Hotel reviewed (approved):
Failed tests seem flaky to me.

for Semantic version failure, we need to raise a JIRA for approval.

engcom-Charlie committed:
Merge branch '2.4-develop' into email-transport-builder-attachments

engcom-Charlie commented:
@magento run all tests

engcom-Charlie commented:
Hi @dudzio12,

Thank you for your contribution!

### :heavy_check_mark: QA Passed

In order to proceed further on this PR, I have followed the below steps on 2.4-develop.
1. addAttachment feature is not avaiable on magento 2.4-develop.  Created a script for sending email with attachment. Got error on addAttachment()

**Before:** :heavy_multiplication_x: 
<img width="1303" alt="image" src="https://github.com/magento/magento2/assets/51680850/9a7c075c-8628-4689-8ce7-d7340966605d">


**After:** :heavy_check_mark:  
Email sent successfully with attachment.
<img width="1168" alt="image" src="https://github.com/magento/magento2/assets/51680850/b2ee3113-be22-49de-88c6-6c912be81ac5">


Thank you!

engcom-Charlie commented:
Hi @dudzio12,

Thank you for your contribution. We have discussed with Product Owner on this feature request. 

The changes in this PR are BIC. As addAttachment can be achievable by other different ways, we can try to avoid these changes in framework classes hence closing this PR now.

Thank you!

ImanuelBertrand commented:
Hi @engcom-Charlie ,

I don't quite understand how this is BIC. I my understanding, this PR adds a public method "addAttachment" to TransportBuilder.
Calling this method is optional, i.e. old code should work as before. Can you explain?

Thank you and best regards,
Imanuel

barryvdh commented:
I understand that there are workaround for this, but imho this seems like a change that provides benefit to the users without impact. Otherwise you have multiple modules adding their own addAttachment methods.

dudzio12 commented:
It's just sad, that it was basically ready to merge and get rejected

ImanuelBertrand commented:
Just for the record, since the decision by Magento is pretty intransparent without any further explanation: Can you, as the author, confirm that there are no backwards incompatibel changes in the PR?

dudzio12 commented:
@engcom-Charlie Could you review this PR one more time? Please explain why you marked this PR as backward incompatible.

engcom-Hotel committed:
Merge branch '2.4-develop' into email-transport-builder-attachments

engcom-Hotel committed:
Merge branch '2.4-develop' into email-transport-builder-attachments

engcom-Hotel commented:
@magento run all tests

engcom-Hotel reviewed (changes requested):
Hello @dudzio12,

Thanks for the contribution!

There are multiple test failures in this PR, please resolve them so that we can proceed further with this PR.

Thanks

engcom-Charlie committed:
Merge branch '2.4-develop' into email-transport-builder-attachments

engcom-Charlie commented:
> > @engcom-Charlie:
> > > The developer’s class extends the core class and relies on the absence of addAttachment() in the parent.
> > 
> > 
> > What parent are you guys talking about? I also don't see this class (`TransportBuilder`) extending from another class, but maybe I'm missing something?
> 
> I think he means that if a developer is extending this class, and the method already exists with a different signature, it will cause an exception.

We are talking about the scenario where, an extension developer, extends TransportBuilder class to his own child class.

lamskoy commented:
Hello community!
Have made extension which fixes issue with email attachments in elegant way.
Have published it on packagist.org 
It can be installed with composer:
```
composer require elcommerce/magento2-advanced-mailer
```
